api/v1beta2/foundationdbcluster_types.go

```diff
@@ -361,18 +362,29 @@ func (processGroupStatus *ProcessGroupStatus) MarkForRemoval() {
 }
 
 // NeedsReplacement checks if the ProcessGroupStatus has conditions so that it should be removed
-func (processGroupStatus *ProcessGroupStatus) NeedsReplacement(failureTime int) (bool, int64) {
-	var missingTime *int64
-	for _, condition := range conditionsThatNeedReplacement {
-		conditionTime := processGroupStatus.GetConditionTime(condition)
-		if conditionTime != nil && (missingTime == nil || *missingTime > *conditionTime) {
-			missingTime = conditionTime
+func (processGroupStatus *ProcessGroupStatus) NeedsReplacement(failureTime int, taintReplacementTime int) (bool, int64) {
+	var earliestFailureTime *int64
+	var earliestTaintReplacementTime *int64
+	for _, conditionType := range conditionsThatNeedReplacement {
+		conditionTime := processGroupStatus.GetConditionTime(conditionType)
+		if conditionType != NodeTaintReplacing {
+			if conditionTime != nil && (earliestFailureTime == nil || *earliestFailureTime > *conditionTime) {
+				earliestFailureTime = conditionTime
+			}
+		} else {
+			if conditionTime != nil && (earliestTaintReplacementTime == nil || *earliestTaintReplacementTime > *conditionTime) {
+				earliestTaintReplacementTime = conditionTime
+			}
```

Comment:
One proposal from my side to remove the some of the complexity in this method:


```go
// NeedsReplacement checks if the ProcessGroupStatus has conditions so that it should be removed
func (processGroupStatus *ProcessGroupStatus) NeedsReplacement(failureTime int, taintReplacementTime int) (bool, int64) {
	if processGroupStatus.IsMarkedForRemoval() {
		return false, 0
	}

	var earliestFailureTime int64 = math.MaxInt64
	var earliestTaintReplacementTime int64 = math.MaxInt64

	for _, conditionType := range conditionsThatNeedReplacement {
		conditionTimePtr := processGroupStatus.GetConditionTime(conditionType)
		if conditionTimePtr == nil {
			continue
		}

		conditionTime := *conditionTimePtr

		if conditionType == NodeTaintReplacing {
			if earliestTaintReplacementTime > conditionTime {
				earliestTaintReplacementTime = conditionTime
			}
			continue
		}

		if earliestFailureTime > conditionTime {
			earliestFailureTime = conditionTime
		}
	}

	failureWindowStart := time.Now().Add(-1 * time.Duration(failureTime) * time.Second).Unix()
	if earliestFailureTime < failureWindowStart {
		return true, earliestFailureTime
	}

	taintWindowStart := time.Now().Add(-1 * time.Duration(taintReplacementTime) * time.Second).Unix()
	if earliestTaintReplacementTime < taintWindowStart {
		return true, earliestTaintReplacementTime
	}

	return false, 0
}
````